### PR TITLE
Improved filters

### PR DESCRIPTION
**Pull request by flimzy**
This PR makes filter writing much easier for humans, by allowing extra whitespace in the filter input.

As an example, rather than writing:

    filter = "(&(mail=*)(|(objectClass=contact)(objectClass=group)(objectClass=inetOrgPerson)(objectClass=user)(objectClass=publicFolder))(!(msExchUserAccountControl:1.2.840.113556.1.4.803:=2))(|(userAccountControl:1.2.840.113556.1.4.803:=256)(userAccountControl:1.2.840.113556.1.4.803:=512)(!(userAccountControl=*)))(!(&(msExchWhenMailboxCreated=*)(!(homeMDB=*)))))"

It is now possible to specify the same filter in this much more readable format:

        filter =  `
            (& (mail=*)
                    (|  (objectClass=contact)
                        (objectClass=group)
                        (objectClass=inetOrgPerson)
                        (objectClass=user)
                        (objectClass=publicFolder)
                    )
                    (!  (msExchUserAccountControl:1.2.840.113556.1.4.803:=2))
                    (|  (userAccountControl:1.2.840.113556.1.4.803:=256)
                        (userAccountControl:1.2.840.113556.1.4.803:=512)
                        (!(userAccountControl=*))
                    )
                    (!  (&  (msExchWhenMailboxCreated=*)
                            (!(homeMDB=*))
                        )
                    )
                )
            `

### CODE REVIEW

**Comment by johnweldon** [comment]
Hmm; the filter tests are failing:
```
08:14:49 [0]$ go test ./...
--- FAIL: TestFilter (0.00s)
        filter_test.go:46: "(sn=*i*le*)" expected, got "(sn=*i*)"
        filter_test.go:46: "(sn=Mi*l*r)" expected, got "(sn=Mi*)"
        filter_test.go:46: "(sn=Mi*le*)" expected, got "(sn=Mi*)"
        filter_test.go:46: "(sn=*i*ler)" expected, got "(sn=*i*)"

```

Otherwise this PR looks interesting.

**Comment by flimzy** [comment]
I just noticed the same, and it appears to me that the old tests were actually broken.  I've updated the PR.  If my diagnosis appears to be wrong along those lines, please let me know.

**Comment by johnweldon** [comment]
@flimzy your comments may be correct about the RFC intended behaviour.  I believe the interpretation you removed was added by @niemeyer who I trust fairly highly.

I'll take some time to review, and decide the right approach.

Thanks for your work on this!

File: filter.go [hunk 1]
```diff
@@ -50,68 +101,328 @@ var FilterSubstringsMap = map[uint64]string{
...
+				fmt.Println("any : " + matches[1])
+			}
+			seq.AppendChild(ber.NewString(ber.ClassContext, ber.TypePrimitive, FilterSubstringsAny, UnescapeFilterValue(matches[1]), "any"))
+		}
+
+		pos += len(matches[0])
+		if pos == len(value) {
+			break
+		}
+	}
```
Comment (flimzy) [inline filter.go]: The above for loop is what should be handling these compound substring matches. It appears that it treasts (sn=*foo*bar*) as multiple substring matches (which I think is correct), where as previously it was treated as a single one.

If my understanding is correct, perhaps the removed tests should be replaced with some tests that verify that the substring match is properly broken into multple substring segments?

**Comment by niemeyer** [comment]
FWIW, I have code in production that depends on these filters actually working, and the application logic that depends on that was ported from my previous Erlang application which also worked fine, and both of these applications with such wildcard matching logic work correctly, against a real LDAP server.

So, if you are going to modify that logic, may I please ask that the changes be confirmed against a real LDAP server?  If you are just disabling the tests, you are most likely breaking real working code.

**Comment by flimzy** [comment]
I have done a simple pragmatic test, using both parsing algorithms, with the following search:

    (mail=*box*dd*6c-*)

And both work fine.  That surprises me. If the previous code was working, I would expect my version to break things.  Perhaps Exchange is smart enough to handle both formats.

In any case, I'll modify my patch to use the old logic, since we know it's working.

**Comment by niemeyer** [comment]
A single match still sounds ideal, even if both work for you. It's also not clear if both are actually working for you or if the test simply did not hit the right case. Note that "*i*le*" is not the same as "*i*" and "*le*". The latter matches "leroi", the former does not.

Either way, thanks for not modifying the currently working code.

**Comment by flimzy** [comment]
@niemeyer, My original analysis was incomplete, and therefore likely incorrect.  It appears there was actually a bug in mavericknz's version of DecompileFilter(). I have reverted that (it wasn't the reason I wanted this patch anyway).  This new version ought to pass all existing tests.

**Comment by johnweldon** [comment]
@flimzy Thanks again.

This change looks like you put some thought into it; I'm not sure how much is your original work and how much comes from @mavericknz, but it's evident that much attention has been given to build on the existing work in Perl, with reference to the RFC.


I have a couple reservations here.

1) The advertised benefit of this PR is increased readability and ease of filter writing. While that is a worthy benefit, doing it by rewriting a significant portion of the filter parsing causes the change to have a disproportionate amount of risk of introducing new subtle defects.

2) I'm not sure how important it is, but switching from the current character by character parsing to using regular expressions for parsing seems like it would have an impact on the performance of the filter parsing.  Again, I don't think this operation happens typically in tight loops or repeatedly enough for this to be a problem, but on the other hand maybe it does.

3) Finally, with the high ratio of new code, I would be much more comfortable with many more tests, especially ones subjecting the regex parsing to many edge case variations.  The tests could also double as documentation for expected inputs on the new functions.

Seem reasonable?

**Comment by flimzy** [comment]
Thank you for your consideration.  My time into this was minimal.  If you choose to reject the PR for any reason, it won't be a big waste of my time--and I may work on another patch of my own to add similar functionality to your existing code.

I will attempt to address some of your concerns, but not directly in order, but rather in the order of ease of testing them. :)

First, on the issue of timing.  I ran some tests on the following filters (using Sprintf to assure there was no sort of caching going on anywhere--I don't know if that's actually a possible concern, but it's easy to avoid) looping the filter generation 1,000,000 in a tight loop.

1. fmt.Sprintf("(mail=%d)",i)

 Old code: 3.949s
 Regex code: 3.818s

2. fmt.Sprintf("(&(mail=%d)(|(objectClass=contact)(objectClass=group)(objectClass=inetOrgPerson)(objectClass=user)(objectClass=publicFolder))(!(msExchUserAccountControl:1.2.840.113556.1.4.803:=2))(|(userAccountControl:1.2.840.113556.1.4.803:=256)(userAccountControl:1.2.840.113556.1.4.803:=512)(!(userAccountControl=*)))(!(&(msExchWhenMailboxCreated=*)(!(homeMDB=*)))))",i)

 Old code: 4.728s
 Regex code: 4.730s

3. The human-readable spaced-out version of #2 (as seen in the original PR description above)

 Old code: n/a
 Regex code: 7.102s

So it seems (somewhat surprisingly to me) that the regex version, even for a somewhat involved filter, performs practically identically to the original version. That's not to say that my sample filters are necessarily representative of most real-world cases. If you think further tests are called for, I'm happy to oblige, especially if you can offer suggested filters to test against.

**Comment by flimzy** [comment]
Regarding your concern #1, I agree.  There is a certain risk involved.  Mavericknz's repo has a fair number of stars on GitHub (although only 3 followers).  It seems to have a fair number of people use it (based on my completely unscientific impression based primarily on it being mentioned on stackoverflow). This is just to say that his code isn't completely green.  But it's also not completely unaltered, and it's possible I've introduced bugs even in my work.

And finally, regarding concern #3, I don't mind working on adding new tests.  If concern #1 is great enough, my time might be better spent, however, working on adding whitespace-removal to the existing code instead?  Thoughts?

**Comment by johnweldon** [comment]
@flimzy thanks again.

I think that a simpler patch just adding whitespace removal to the existing code would be easier to vet and merge quickly.

After that if there is still justification for the regexp changes then we should add tests around the new code.

As far as performance testing, I wasn't clear from your comments if you used the Go benchmarking mechanism - you may be interested in this link: http://dave.cheney.net/2013/06/30/how-to-write-benchmarks-in-go

**Comment by flimzy** [comment]
I have moved to a new project, and won't be doing any Go development in the foreseeable future, so I won't be able to produce the alternate PR I was planning to work on.

Feel free to close this PR if you wish, or if someone else wants to take it over, that's fine with me, too.

**Comment by johnweldon** [comment]
Thanks @flimzy hopefully we can pick this up some time again.
